### DISCUSSION TITLE
Issue with Narration in Portuguese - Trailing Dots

### DISCUSSION THREAD

**Discussion by arcanjotc** [erew123/alltalk_tts#352] [Q&A] [answered]
Hello AllTalk Community,

I’m having an issue with the narration feature in Portuguese. Even though I have the TG-webui Remove Trailing Dots: true option enabled, the narrator still reads the trailing dots at the end of sentences.

Has anyone experienced this or have any suggestions on how to fix it? I would appreciate any help!

Thank you!

**Comment by erew123** [comment]
Hi @arcanjotc 

Im guessing you are on AllTalk v1. Dots/periods, at least as far as Latin character set ones, should already be removed forcefully by this line:

https://github.com/erew123/alltalk_tts/blob/main/script.py#L581

Its possible there could be Portuguese specific characters sets that are in use, such as:

Standard ASCII period (.) - U+002E
Full stop (．) - U+FF0E
One dot leader (․) - U+2024
Middle dot (·) - U+00B7
Bullet (•) - U+2022

Im not sure if its one of those you may be seeing.

You could edit those lines in your `script.py` file 

```
    # Replace ellipsis with a single dot
    text = re.sub(r"\.{3,}", ".", text)
```

and replace it with:

```    
    # Define a set of characters considered as dots
    dots = r'[.．․·•]'

    # Replace ellipsis (three or more dot characters) with a single dot
    text = re.sub(f'{dots}{{3,}}', '.', text)

    # Remove trailing dots if the checkbox is checked
    if params["remove_trailing_dots"]:
        # Split the text into sentences
        sentences = re.split(r'(?<=[.!?．。！？])\s+', text)
        # Remove trailing dots from each sentence
        sentences = [re.sub(f'{dots}+$', '', s) for s in sentences]
        # Join the sentences back together
        text = ' '.join(sentences)
```

Which is a far more aggressive variation of removing trailing dots. 

It just really comes back to what the dots actually are. Also, I am assuming you are seeing them at the AllTalk console output as being part of the TTS generation. They would NOT be removed from what you see on screen in Text-Gen-Webui, they would be stripped when they are passed over to AllTalk and would not be visible in AllTalk's command line output.

Thanks

**Reply by arcanjotc** [reply]
Yes. I'm seeing them at the AllTalk console output as being part of the TTS generation.

**Reply by arcanjotc** [reply]
> Hi @arcanjotc
> 
> Im guessing you are on AllTalk v1. Dots/periods, at least as far as Latin character set ones, should already be removed forcefully by this line:
> 
> https://github.com/erew123/alltalk_tts/blob/main/script.py#L581
> 
> Its possible there could be Portuguese specific characters sets that are in use, such as:
> 
> Standard ASCII period (.) - U+002E Full stop (．) - U+FF0E One dot leader (․) - U+2024 Middle dot (·) - U+00B7 Bullet (•) - U+2022
> 
> Im not sure if its one of those you may be seeing.
> 
> You could edit those lines in your `script.py` file
> 
> ```
>     # Replace ellipsis with a single dot
>     text = re.sub(r"\.{3,}", ".", text)
> ```
> 
> and replace it with:
> 
> ```
>     # Define a set of characters considered as dots
>     dots = r'[.．․·•]'
> 
>     # Replace ellipsis (three or more dot characters) with a single dot
>     text = re.sub(f'{dots}{{3,}}', '.', text)
> 
>     # Remove trailing dots if the checkbox is checked
>     if params["remove_trailing_dots"]:
>         # Split the text into sentences
>         sentences = re.split(r'(?<=[.!?．。！？])\s+', text)
>         # Remove trailing dots from each sentence
>         sentences = [re.sub(f'{dots}+$', '', s) for s in sentences]
>         # Join the sentences back together
>         text = ' '.join(sentences)
> ```
> 
> Which is a far more aggressive variation of removing trailing dots.
> 
> It just really comes back to what the dots actually are. Also, I am assuming you are seeing them at the AllTalk console output as being part of the TTS generation. They would NOT be removed from what you see on screen in Text-Gen-Webui, they would be stripped when they are passed over to AllTalk and would not be visible in AllTalk's command line output.
> 
> Thanks

I managed to find and replace the lines of code, but it seems it hasn't worked. It continues to pronounce the dots, and they also appear in the output console.

If you have any other ideas, I would love to try implementing them.

Thank you in advance!

**Reply by erew123** [reply]
@arcanjotc I see why. Sorry I havnt been need V1's code for quite some time now. 

Ok, so you need to replace this block of code in script.py (now youve already changed the code a little, the line numbers may be a little different)

https://github.com/erew123/alltalk_tts/blob/main/script.py#L727-L736

```
 		# Decode HTML entities first
                cleaned_part = html.unescape(cleaned_string)
                # Define a set of characters considered as dots
                dots = r'[.．․·•\u3002]'  # Added \u3002 for ideographic full stop
                # Replace ellipsis (three or more dot characters) with a single dot
                cleaned_part = re.sub(f'{dots}{{3,}}', '.', cleaned_part)
                # Replace multiple instances of certain punctuation marks with a single instance
                cleaned_part = re.sub(r"([!?.\u3002\uFF1F\uFF01\uFF0C])\1+", r"\1", cleaned_part)
                # Replace "Chinese ellipsis" with a single dot
                cleaned_part = re.sub(r"\u2026{1,2}", ". ", cleaned_part)
                # Remove trailing dots if the checkbox is checked
                if params["remove_trailing_dots"]:
                    # Split the text into sentences
                    sentences = re.split(r'(?<=[.!?．。！？])\s+', cleaned_part)
                    # Remove trailing dots from each sentence
                    sentences = [re.sub(f'{dots}+$', '', s) for s in sentences]
                    # Join the sentences back together
                    cleaned_part = ' '.join(sentences)
                # Further clean to remove any other unwanted characters
                cleaned_part = re.sub(r'[^a-zA-Z0-9\s.,;:!?\-\'"$\u0400-\u04FF\u00C0-\u00FF\u0150\u0151\u0170\u0171\u0900-\u097F\u2018\u2019\u201C\u201D\u3001\u3002\u3040-\u309F\u30A0-\u30FF\u4E00-\u9FFF\u3400-\u4DBF\uF900-\uFAFF\u0600-\u06FF\u0750-\u077F\uFB50-\uFDFF\uFE70-\uFEFF\uAC00-\uD7A3\u1100-\u11FF\u3130-\u318F\uFF01\uFF0c\uFF1A\uFF1B\uFF1F]', '', cleaned_part) 
                # Remove all newline characters (single or multiple)
                cleaned_part = re.sub(r"\n+", " ", cleaned_part)
```

So you are replacing the bits between `else` and `# Process the part and give it a non-character name if being used vai API or standalone.` with the above. 

If you search for `# Decode HTML entities first` that should take you to the correct line. 

Thanks

**Reply by arcanjotc** [reply]
> @arcanjotc I see why. Sorry I havnt been need V1's code for quite some time now.
> 
> Ok, so you need to replace this block of code in script.py (now youve already changed the code a little, the line numbers may be a little different)
> 
> https://github.com/erew123/alltalk_tts/blob/main/script.py#L727-L736
> 
> ```
>  		# Decode HTML entities first
>                 cleaned_part = html.unescape(cleaned_string)
>                 # Define a set of characters considered as dots
>                 dots = r'[.．․·•\u3002]'  # Added \u3002 for ideographic full stop
>                 # Replace ellipsis (three or more dot characters) with a single dot
>                 cleaned_part = re.sub(f'{dots}{{3,}}', '.', cleaned_part)
>                 # Replace multiple instances of certain punctuation marks with a single instance
>                 cleaned_part = re.sub(r"([!?.\u3002\uFF1F\uFF01\uFF0C])\1+", r"\1", cleaned_part)
>                 # Replace "Chinese ellipsis" with a single dot
>                 cleaned_part = re.sub(r"\u2026{1,2}", ". ", cleaned_part)
>                 # Remove trailing dots if the checkbox is checked
>                 if params["remove_trailing_dots"]:
>                     # Split the text into sentences
>                     sentences = re.split(r'(?<=[.!?．。！？])\s+', cleaned_part)
>                     # Remove trailing dots from each sentence
>                     sentences = [re.sub(f'{dots}+$', '', s) for s in sentences]
>                     # Join the sentences back together
>                     cleaned_part = ' '.join(sentences)
>                 # Further clean to remove any other unwanted characters
>                 cleaned_part = re.sub(r'[^a-zA-Z0-9\s.,;:!?\-\'"$\u0400-\u04FF\u00C0-\u00FF\u0150\u0151\u0170\u0171\u0900-\u097F\u2018\u2019\u201C\u201D\u3001\u3002\u3040-\u309F\u30A0-\u30FF\u4E00-\u9FFF\u3400-\u4DBF\uF900-\uFAFF\u0600-\u06FF\u0750-\u077F\uFB50-\uFDFF\uFE70-\uFEFF\uAC00-\uD7A3\u1100-\u11FF\u3130-\u318F\uFF01\uFF0c\uFF1A\uFF1B\uFF1F]', '', cleaned_part) 
>                 # Remove all newline characters (single or multiple)
>                 cleaned_part = re.sub(r"\n+", " ", cleaned_part)
> ```
> 
> So you are replacing the bits between `else` and `# Process the part and give it a non-character name if being used vai API or standalone.` with the above.
> 
> If you search for `# Decode HTML entities first` that should take you to the correct line.
> 
> Thanks

I appreciate your help and I really didn't mean to bother you.

I replaced the block, but it remains the same.
Could there be some cache that needs to be cleared or something like that?
As can be seen in the image, it still shows the dot:
![Capturar](https://github.com/user-attachments/assets/63856e27-a007-4595-af1c-3162a1a26fb8)

I made a way that works perfectly, but I need to manually remove all the periods and place a dash at the beginning of each sentence to create a pause. As can be seen in the image.
![Capturar2](https://github.com/user-attachments/assets/80a4f50e-e5e6-418c-acfc-35eff4e78934)

**Reply by erew123** [reply, marked as answer]
To do that would be completely custom code to do something like that....

```
# Remove all newline characters (single or multiple)
cleaned_part = re.sub(r"\n+", " ", cleaned_part)
# New code for replacing periods with dashes
# Split the text into sentences
sentences = re.split(r'(?<=[.!?．。！？])\s+', cleaned_part)
processed_sentences = []
for i, sentence in enumerate(sentences):
    # Remove leading and trailing whitespace
    sentence = sentence.strip()    
    # Remove the period at the end if present
    if sentence.endswith('.'):
        sentence = sentence[:-1]
    # Add a dash before the sentence, but not for the first sentence
    if i > 0:
        sentence = '- ' + sentence
    processed_sentences.append(sentence)
# Join the sentences back together
cleaned_part = ' '.join(processed_sentences)
# Clean up any double spaces that might have been created
cleaned_part = re.sub(r'\s+', ' ', cleaned_part)
# Ensure there's no space between consecutive dashes (in case of empty sentences)
cleaned_part = re.sub(r'- +- +', '- ', cleaned_part)
```

**Input:** "Hello world. This is a test. How are you?"
**Output:** "Hello world - This is a test - How are you"

You're welcome to play around with that and try it after the:

```
                # Decode HTML entities first
                cleaned_part = html.unescape(cleaned_string)
```

but, you're in the realms of custom code as I say and doing something that's unique to your own situation, so you may need to look at the code and get it to work for your scenarios.

Thanks

**Reply by arcanjotc** [reply]
> # Remove all newline characters (single or multiple)
> cleaned_part = re.sub(r"\n+", " ", cleaned_part)
> # New code for replacing periods with dashes
> # Split the text into sentences
> sentences = re.split(r'(?<=[.!?．。！？])\s+', cleaned_part)
> processed_sentences = []
> for i, sentence in enumerate(sentences):
>     # Remove leading and trailing whitespace
>     sentence = sentence.strip()    
>     # Remove the period at the end if present
>     if sentence.endswith('.'):
>         sentence = sentence[:-1]
>     # Add a dash before the sentence, but not for the first sentence
>     if i > 0:
>         sentence = '- ' + sentence
>     processed_sentences.append(sentence)
> # Join the sentences back together
> cleaned_part = ' '.join(processed_sentences)
> # Clean up any double spaces that might have been created
> cleaned_part = re.sub(r'\s+', ' ', cleaned_part)
> # Ensure there's no space between consecutive dashes (in case of empty sentences)
> cleaned_part = re.sub(r'- +- +', '- ', cleaned_part)

I understand and really appreciate your help.

I’ll play around with the code a bit and see what I can do.

In any case, it would be interesting to implement this feature in future versions. In the tests I conducted, there are other languages that also face this issue, like Italian, for example.

I’ll keep using Alltalk for English and will try to modify the code to solve the issue with the dots.

Without a doubt, Alltalk is the best alternative at the moment!

Once again, I appreciate your help.
Thank you very much.

**Reply by pericron** [reply]
I also use AllTalk in Portuguese and face the same problem. I'd appreciate this implemented natively as well. Perhaps I'll open an Issue to help out.

**Reply by arcanjotc** [reply]
What I have been doing to minimize the issue is replacing all the "." to ":"  before generating. This action has made Alltalk usable in Brazilian Portuguese. It takes a little bit of work, but it works.

**Reply by erew123** [reply]
Hi @GreyGoldFish @arcanjotc I am away as you may have seen in other comments on here, but I have put out a statement on [this here](https://github.com/erew123/alltalk_tts/issues/377)

I will look at things down the line. If you have a specific "when generating as Portuguese X character shoud be converted to X character" I should be able to do something, but you would need to let me know exactly what works... OR... if it is specific to the Coqui XTTS model/engine, you can always mention it to Idiap, whom is maintaining the Coqui TTS engines and their tokenisers and they should be able to build it into the Coqui TTS engine. The link to their Github is here https://github.com/idiap/coqui-ai-TTS

Thanks
